python/gstaichi/lang/_fast_caching/args_hasher.py

```diff
@@ -94,14 +112,19 @@ def stringify_obj_type(path: tuple[str, ...], obj: object) -> str | None:
     return None
 
 
-def hash_args(args: Sequence[Any]) -> str | None:
+def hash_args(args: Sequence[Any], arg_metas: Sequence[ArgMetadata | None]) -> str | None:
     global g_num_calls, g_num_args, g_hashing_time, g_repr_time, g_num_ignored_calls
     g_num_calls += 1
     g_num_args += len(args)
     hash_l = []
+    if len(args) != len(arg_metas):
+        raise RuntimeError(
+            f"Number of args passed in {len(args)} doesnt match number of declared args {len(arg_metas)}"
+        )
     for i_arg, arg in enumerate(args):
         start = time.time()
-        _hash = stringify_obj_type((str(i_arg),), arg)
+        arg_meta = arg_metas[i_arg]
+        _hash = stringify_obj_type((str(i_arg),), arg, arg_meta)
```

Comment:
`_hash = stringify_obj_type((str(i_arg),), arg, arg_metas[i_arg])` is good enough I think. But as you want.